cli/command_repository_set_parameters_test.go

```diff
@@ -182,6 +182,71 @@ func (s *formatSpecificTestSuite) TestRepositorySetParametersUpgrade(t *testing.
 	env.RunAndExpectSuccess(t, "index", "epoch", "list")
 }
 
+func (s *formatSpecificTestSuite) TestRepositorySetParametersDowngrade(t *testing.T) {
+	env := s.setupInMemoryRepo(t)
+	out := env.RunAndExpectSuccess(t, "repository", "status")
+
+	// default values
+	require.Contains(t, out, "Max pack length:     21 MB")
+
+	switch s.formatVersion {
+	case format.FormatVersion1:
+		require.Contains(t, out, "Format version:      1")
+		require.Contains(t, out, "Epoch Manager:       disabled")
+		env.RunAndExpectFailure(t, "index", "epoch", "list")
+	case format.FormatVersion2:
+		require.Contains(t, out, "Format version:      2")
+		require.Contains(t, out, "Epoch Manager:       enabled")
+		env.RunAndExpectSuccess(t, "index", "epoch", "list")
+	default:
+		require.Contains(t, out, "Format version:      3")
+		require.Contains(t, out, "Epoch Manager:       enabled")
+		env.RunAndExpectSuccess(t, "index", "epoch", "list")
+	}
+
+	env.Environment["KOPIA_UPGRADE_LOCK_ENABLED"] = "1"
+
+	{
+		cmd := []string{
+			"repository", "upgrade",
+			"--upgrade-owner-id", "owner",
+			"--io-drain-timeout", "1s", "--allow-unsafe-upgrade",
+			"--status-poll-interval", "1s",
+			"--max-permitted-clock-drift", "1s",
+		}
+
+		// You can only upgrade when you are not already upgraded
+		if s.formatVersion < format.MaxFormatVersion {
+			env.RunAndExpectSuccess(t, cmd...)
+		} else {
+			env.RunAndExpectFailure(t, cmd...)
+		}
+	}
+
+	env.RunAndExpectSuccess(t, "repository", "set-parameters", "--upgrade")
+	env.RunAndExpectSuccess(t, "repository", "set-parameters", "--epoch-min-duration", "3h")
+	env.RunAndExpectSuccess(t, "repository", "set-parameters", "--epoch-cleanup-safety-margin", "23h")
+	env.RunAndExpectSuccess(t, "repository", "set-parameters", "--epoch-advance-on-size-mb", "77")
+	env.RunAndExpectSuccess(t, "repository", "set-parameters", "--epoch-advance-on-count", "22")
+	env.RunAndExpectSuccess(t, "repository", "set-parameters", "--epoch-checkpoint-frequency", "9")
+
+	env.RunAndExpectFailure(t, "repository", "set-parameters", "--epoch-min-duration", "1s")
+	env.RunAndExpectFailure(t, "repository", "set-parameters", "--epoch-refresh-frequency", "10h")
+	env.RunAndExpectFailure(t, "repository", "set-parameters", "--epoch-checkpoint-frequency", "-10")
+	env.RunAndExpectFailure(t, "repository", "set-parameters", "--epoch-cleanup-safety-margin", "10s")
+	env.RunAndExpectFailure(t, "repository", "set-parameters", "--epoch-advance-on-count", "1")
+
+	out = env.RunAndExpectSuccess(t, "repository", "status")
+	require.Contains(t, out, "Epoch Manager:       enabled")
+	require.Contains(t, out, "Index Format:        v2")
+	require.Contains(t, out, "Format version:      3")
+	require.Contains(t, out, "Epoch cleanup margin:    23h0m0s")
+	require.Contains(t, out, "Epoch advance on:        22 blobs or 80.7 MB, minimum 3h0m0s")
+	require.Contains(t, out, "Epoch checkpoint every:  9 epochs")
```

Comment:
same suggestion ☝️  for `require.Regexp` would apply here

Reply:
@redgoat650 `require.Contains` is part of the testify framework, so changing the implementation of `Contains` will be challenging.  Using, or creating, another assertion that would allow regular expressions is a good approach.  The rest of the code that inspects output from `require.Contains` should be moved over as well (otherwise there will be one bit of code that doesn't conform).

(I added ticket fix this for the kasten branch)

Reply:
@aaron-kasten - I wasn't suggesting changing the behavior of `require.Contains`. I was suggesting you instead use  the existing `require.Regexp` instead. Both are part of the testify framework.

https://github.com/stretchr/testify/blob/1333b5d3bda8cf5aedcf3e1aaa95cac28aaab892/require/require.go#L173